FEATURE: New link editor (Archaeopteryx)

mhsdesign opened the pull request:
<!--
Thanks for your contribution, we appreciate it!

ATTENTION: ALL NEW FEATURE PRs SHOULD TARGET THE MASTER BRANCH (bugfixes go to the least maintained branch)
-->


The package is herby merged into the Neos Ui from https://github.com/sitegeist/Sitegeist.Archaeopteryx/tree/2.0.1

- [Initial working adaption of Archaeopteryx](https://github.com/neos/neos-ui/commit/a5ef804eab0ebe0332ca37a6919255be13dd332c)
- ...

With these adjustments the copied code already works and offers the new link editor experience though the task is not complete yet:

- [ ] add link editor documentation to neos docs or some place -> document that `anchor: true` is obsolete and that first level starting point is deprecated.
- [x] remove previous ckeditor link editor and inspector link editor
- [x] remove php linting from circle ci in favour of new workflow -> https://github.com/neos/neos-ui/issues/4026
- [x] discuss where to place `Neos\Neos\Ui\LinkEditor\Link` DTO (the thing which is stuffed into the Node) -> https://github.com/neos/neos-development-collection/pull/5682
- [x] fix `LinkToArrayForNeosUiConverter` hack in core -> https://github.com/neos/neos-ui/pull/3723
- [x] remove neos-ui-link-editor-bridge corruption layer
- [x] remove styled-components and use css modules
- [x] remove rx-js and use framework-observeable if possible
- [x] evaluate the need for react-final-form and final-form (all other neos ui parts are coded without)
- [x] remove or combine neos-ui-link-editor-error-handling in favour of our own error handling
- [x] commit all new .yarn dependencies but evaluate the necessity of each
- [x] satisfy js linting and typescript
- [x] optimise bundle size by lazy loading the huge library "libphonenumber-js"
- [x] ~document new plain node-js tests and also see that we only use one version of esbuild for the testing too (evaluate if we can write node-js tests in typescript?)~
- [x] reintroduce `download` attribute support - ui feature since https://github.com/neos/neos-ui/pull/3197
- [x] extract inline styles to css modules
- [x] extract neos globals change and evaluate where to place the global registry and config
  - https://github.com/neos/neos-ui/pull/4010
  - https://github.com/neos/neos-ui/pull/4011
  - https://github.com/neos/neos-ui/pull/4012
- [x] unify icon choices e.g. use cross (x) icon instead of bin like in assets editor
- [x] review how to use the new linktype configuration
- [x] paulas review
- [x] vertical tabs
- [x] do not use `action: ingored` maybe fix flow already and build my and christians idea or use `FlowRoute` annotation?
- [x] labels should not follow a double colon
- [x] link string must fully hide options icon
- [x] add proper translations source for all warning texts in tabs
- [x] Phone Number to Phone (also in translations)
- [x] review new components in `presentation` if they should be react ui components -> in the future add a shared components package for these internals
- [x] review introduces new packages, or if things should be grouped `neos-ui-link-editor-core`, `neos-ui-link-editor-custom-node-tree`,  `neos-ui-link-editor-inspector-editor`, `neos-ui-link-editor-link-button`

Optional things for the future
- [ ] optional: introduce disabling of some link options? Disable per tab via yaml or that the linktype only accepts certain options
- [ ] optional: outline and show future idea of `rel` select box with custom entries from redactor
- [ ] optional: hover effect in inspector preview looks odd for assets as it makes them really bright
- [ ] optional: Link Preview is missing in inline editor (and also a way to goto the selected target?)
- [ ] optional: fix absolute height hack for node and asset link targets
- [ ] optional: reintroduce Ctrl+K LinkEditor shortcut - ui feature since https://github.com/neos/neos-ui/pull/2900 -> it was hacky implemented maybe solvable when we build a native ck link plugin


**What I did**

**How I did it**

**How to verify it**

<!--
If possible, a screenshot or a gif comparing the new and old behavior would be great.
-->

mhsdesign committed:
FEATURE: Introduce new LinkEditor based on Sitegeist.Archaeopteryx

The package is herby merged into the Neos Ui from https://github.com/sitegeist/Sitegeist.Archaeopteryx/tree/2.0.1

Following initial changes have been carried out:

- Introduce new github based "server" workflow to run behaviour tests (this should replace all server side linting now done via circleci)
- add all php Classes from "Sitegeist.Archaeopteryx/Classes" to "Classes/LinkEditor" and adapt namespace
- Add policy and backend auth configuration for new endpoints inside LinkEditor
- add route definitions (previously neos/archaeopteryx/ now neos/link-editor/)
- add translations previously in LinkType and Main -> now located in a separate LinkEditor namespace which is fully included
- replace Archaeopteryx `plugin` package via manifest.link-editor.js
- rename Archaeopteryx `core` package to neos-ui-link-editor-core as well as all other (neos-bridge is to be removed)
- add behaviour test for endpoint
- run new node based test via `node --test` and deprecate our jest tests

With these adjustments the copied code already works and offers the new link editor experience though the task is not complete yet:

- add link editor documentation to neos docs or some place
- remove previous ckeditor link editor and inspector link editor
- remove php linting from circle ci in favour of new workflow
- fix LinkToArrayForNeosUiConverter in core
- remove neos-ui-link-editor-bridge corruption layer
- remove styled-components and use css modules
- remove rx-js and use framework-observeable if possible
- evaluate the need for react-final-form and final-form (all other neos ui parts are coded without)
- remove or combine neos-ui-link-editor-error-handling in favour of our own error handling
- commit all new .yarn dependencies but evaluate the necessity of each
- satisfy js linting and typescript
- optimise bundle size by lazy loading the huge library "libphonenumber-js"

mhsdesign committed:
BUGFIX: useLatestState() hook does not setValue if observer emits its default value again

See tests

> initial state, update, set to initial

and

> update state back to defaultValue

This bug becomes eminent when implementing a dialog which is closed by default and then opens - which works - but is later closed again.

mhsdesign committed:
TASK: Remove not implemented `ObservablePipeOperation` (we don't offer `observable$.pipe()`

mhsdesign committed:
TASK: Introduce new ActionObservable in `@neos-project/framework-observable`

This allows to build redux like action state reducing.
The ActionObservable is similar to the rxjs Subject.

mhsdesign committed:
TASK: LinkEditor remove rxjs in favour of our `@neos-project/framework-observable`

mhsdesign committed:
TASK: LinkEditor remove EditorContext and pass IEditor along instead

mhsdesign committed:
pick 106cde283b TASK: Adjust node property mappers to use new node address
pick 0f6284b323 TASK: Adjust `Neos.Node.serializedNodeAddress` to use new node address
pick fa834f3104 fixup! TASK: Adjust controllers to use legacy node address notation explicitly instead of property mapping

TASK: LinkEditor remove corruption NeosContext from "link-editor-neos-bridge"

and defer to using the actual context instead

mhsdesign committed:
TASK: LinkEditor turn all packages private to prevent nmp publish

mhsdesign committed:
TASK: LinkEditor remove corruption useSelector from "link-editor-neos-bridge"

and defer to using react-redux mostly direct

While a direct import from "react-redux" would work, the `useSelector` would not be aware of its state type. And as the defaultState was removed https://github.com/reduxjs/react-redux/issues/1879 we resort to now cleanly reexporting it.

mhsdesign committed:
TASK: LinkEditor remove corruption hooks when selectors are available

mhsdesign committed:
TASK: LinkEditor move partially hacky useSiteNodeAggregateId to Node LinkType for later fixing

mhsdesign committed:
TASK: LinkEditor remove corruption useI18N hook and replace with new translate() function

... if possible ... for user input like options.label the registry is used with its fallback handling.

mhsdesign committed:
TASK: LinkEditor remove corruption useAssetSummary hook and call endpoint directly

mhsdesign committed:
TASK: LinkEditor remove corruption fetchWithErrorHandling layer and use backend connector directly

mhsdesign committed:
TASK: LinkEditor remove corruption useConfiguration and useGlobalRegistry layer and use global state instead

Originally the Neos Ui was build seemingly without real global state. There is a high order component function - invoked via a decorator @neos() which allowed to inject the registry. For functional components a hook could be introduced, or similar to translate() we have a global state.

Fully removes the temporary package @neos-project/neos-ui-link-editor-neos-bridge again

mhsdesign committed:
TASK: Allow generic errorFallback in <ErrorBoundary> ->

- use global translate() function now to avoid passing i18n registry

mhsdesign committed:
TASK: LinkEditor remove `error-handling` in favour of our `@neos-project/neos-ui-error` package

- adjust naming of QueryResponse to be more precise
- return 200 even if there is an error as otherwise fetchWithErrorHandling would do the error handling task and skip the boundary
- centralise isDevelopmentContext via systemEnv in neos-ui
- add class to wrap two errors together as part of a error chain (NestedError)
   this is not an ECMAScript error and thus does not extend Error
   support showing nested errors in the ErrorView
- throw server side error objects directly in javascript
- align link editor server side errors with neos ui (e.g. class, message, code and now optional trace)
- render ErrorView in FatalErrorView as well for precise client side vs server side vs nested error representation
- remove temporary @neos-project/neos-ui-link-editor-error-handling

mhsdesign commented:
## Screenshots (updated 12.09)

### New link editor

<img width="738" height="618" alt="image" src="https://github.com/user-attachments/assets/75d9dad6-d886-4841-aba8-6b71ef7bf8e5" />

### Nested error view from server side errors

<img width="778" height="549" alt="image" src="https://github.com/user-attachments/assets/ea701bf8-5b6b-45cc-b9c6-c7184d9d3802" />

### Improved general fatal error boundary

<img width="1022" height="693" alt="image" src="https://github.com/user-attachments/assets/f39b1718-13fd-435a-a115-1115d88340d6" />

mhsdesign committed:
TASK: Adjust Publishing and Syncing dialog use to non-nullable ErrorView

That makes just not sense to not provide an error :D

mhsdesign committed:
TASK: LinkEditor remove `useKey` dependency by using core's default behaviour

We allow to escape via ESC-key and click outside now if there are no pending changes.
If there are changes we require Cancel to be pressed and shake the dialog with an orange border - just like with the node creation dialog.

mhsdesign committed:
TASK: LinkEditor remove `useDebounce` dependency by using lodash

mhsdesign committed:
TASK: LinkEditor remove `use-async` package dependency by introducing small framework package to handle promises for `usePromise()`

the implementation of usePromise has been taken over from pr #3331 and a rudimentary test has been added.

mhsdesign committed:
TASK: Hotfix remaining react-ui-components to not require `theme` as mandatory prop

technically the definition is correct, but the package `themr()` we use to inject the css styles does not remove the `theme` prop from the mandatory props causing ts error when using the already decorated component without `theme`

im not proud of using ! here but we did so already in other places

mhsdesign committed:
WIP: LinkEditor replace react-final-form with observable from link options fields

mhsdesign committed:
BUGFIX: Declare <IconButton> props as optional when part of `defaultProps`

mhsdesign committed:
TASK: Allow Dialog to carry id (for testing)

mhsdesign committed:
TASK: Correctly set html property id so label for is targeted correctly and to be used for testing

mhsdesign committed:
FEATURE: Add initial e2e NeosUi tests for the LinkEditor

mhsdesign committed:
BUGFIX: Correctly encode mailto url

Replaces https://github.com/sitegeist/Sitegeist.Archaeopteryx/pull/94

Refixes https://github.com/sitegeist/Sitegeist.Archaeopteryx/pull/28 - which complained that the space being encoded to + will not show up correct in the mail app.

Now spaces are correctly encoded as %20 via encodeURIComponent

mhsdesign committed:
TASK: Use `isEmail` helper in link editor to validate input

mhsdesign committed:
TASK: Link Editor Neos ui e2e test for validation, node tree, persistence, link value object and link options

mhsdesign committed:
WIP: LinkEditor replace react-final-form with observable from link options fields

mhsdesign committed:
WIP: LinkEditor replace react-final-form with observable models for each link type

behaviour changes:

- CustomLink now shows proper inline validation errors when text was set to empty string again
- For existing links the next suitable tab is opened (to match the preview of the link)

code cleanups:
- remove react-final form
- remove complex FieldGroups
- remove wrapped NeosEditorEnvelope
- simplify Preview and Editor by not passing `link` only the model
- simplify LinkEditor mounting
- cleanup preview rendering for existing links

dont look into the state and read the link id which would break the abstraction:

> if (!modelFromState || (linkType.id === 'Sitegeist.Archaeopteryx:Web' && !modelFromState?.urlWithoutProtocol) || (linkType.id === 'Sitegeist.Archaeopteryx:PhoneNumber' && !modelFromState?.phoneNumber) || (linkType.id === 'Sitegeist.Archaeopteryx:CustomLink' && !modelFromState?.CustomLink)) {

WIP;
- re-enable other links
- fix unset handling
- fix memory leaks
- fix tab is preselected for new link creation

mhsdesign committed:
WIP: LinkEditor fix reset current tab after closing and pass non-nullable initialLinkType to make ts happy

mhsdesign committed:
WIP: LinkEditor move IProcess framework to new framework-promise-react and name it PromiseState

This will allow to load the big js-phonelib asynchronously

mhsdesign committed:
FEATURE: Enable code splitting for the neos ui by using type:module for Host.js

this is now used to chunk the libphonenumber js library

mhsdesign committed:
WIP: LinkEditor improve error handling (use ErrorView directly)

mhsdesign committed:
WIP: LinkEditor remove html `<form` element to prevent reload on submit

also we dont need it currently as we track the state ourselves

mhsdesign committed:
WIP: LinkEditor reintroduce dirty state handling by moving observeables into root dialog

wip: reset the observerables after closing so reopening works without artefacts

mhsdesign committed:
WIP: LinkEditor simplify unset initial link (only enable the apply button if there was an initial link)

mhsdesign committed:
WIP: LinkEditor reset state after closing by unmounting react component

also fixes the handling of options -> set initial options and apply change

mhsdesign committed:
WIP: LinkEditor migrate Web and MailTo to new observable based state handling in place of final-form

mhsdesign committed:
TASK: LinkEditor remove all leftover "Archaeopteryx" references which leaked into the configuration

The inline configuration changes as the `linkTypes` will now be a primary option in the `linking` settings:

```yaml
'Vendor.Site:MyAwesomeNodeTypeWithSomeRichText':
  # ...
  properties:
    text:
      type: string
      ui:
        inlineEditable: true
        inline:
          editorOptions:
            # ..
            linking:
              #
              # Deprecated, please use the startingPoint option below takes precedence.
              #
              startingPoint: '...'
              linkTypes:
                'LinkEditor:Node':
                  startingPoint: '...' # Default: current site node
                  baseNodeType: 'Vendor.Site' # default: Neos.Neos:Document
```

When uses as inspector editor "Neos.Neos/Inspector/Editors/LinkEditor" must now be used. LinkTypes are still configured via editorOptions,linkTypes.

mhsdesign committed:
PATCH: LinkEditor fix react memory leak in preview

Because of the lazy Tabs ` model = useLatestState(model$)` behaves faulty and returns a mismatching model to the passed `linkType`

react understands the components better if this is moved into a separate `PreviewForLinkType` part.

mhsdesign committed:
TASK: Remove old LinkEditor implementation

- Removes `LinkLookup` data-loader as we now use dedicated endpoints
- Removes the CKEditor LinkButton as this there is now a new button registered in manifest.link-editor.js
- Removes now obsolete "actions.UI.ContentCanvas.toggleLinkEditor"
- Removes now obsolete and shady implemented "actions.UI.Remote.lockPublishing"
- Removes obsolete containerRegistry LinkInput/OptionsPanel/DefaultLinkInputOptions

Todo remove all `Neos.Neos.Ui:Main:ckeditor__toolbar__link` translations

mhsdesign committed:
TASK: Remove unused `NeosAssetLookup` data-loader

introduced with https://github.com/neos/neos-ui/commit/44dc92868601160136232d5ffe2672f8d6f73f10 for node linking

usage removed with https://github.com/neos/neos-ui/commit/e78d99eeb0e4e89ad06e393078c3cffaadab3208 in favour of `AssetLookup`

now that node linking is new implemented we say goodbye

mhsdesign committed:
TASK: Remove barely used `selectors.UI.NodeLinking.contextForNodeLinking` selector

The whole `NodeLinking` module exposes just a combined selector and not own redux state. This removes that complexity.

Also with Neos 9 we want to work more with the node address and its parts ... the old shape workspaceName, contextNode, dimensions should be phased out. Especially as `contextNode` being a json node address contains all information already.

mhsdesign committed:
FEATURE: Reintroduce `download` attribute support for new LinkEditor

based on https://github.com/neos/neos-ui/pull/3197

mhsdesign committed:
TASK: New link editor polish translations

adds placeholders for options
remove translations for old link editor
rename `rel="nofollow"` to `No follow (SEO)`

also removes all LinkType validation translations which are not used any more

mhsdesign committed:
TASK: LinkEditor use neos-ui Tabs component and fix tab opening glitch

previously the Document tab was always loaded first followed by an immediate state update. This is now done initially to avoid any zero-tick component state and unnecessary http requests.

include e2e test to validate value upcasting and that the proper tab is focussed.

mhsdesign committed:
FEATURE: LinkEditor allow `anchor` on Assets to target a specific page of a pdf for example (removes hacky custom link)

This has been asked for in the Archeopteryx package which actually resulted in the introduction of the CustomLink: https://github.com/sitegeist/Sitegeist.Archaeopteryx/commit/0ba0940ad69113b187374c85c684d491eba814f9

Removes the `CustomLink` tab because this is a too wide escape hatch and allows values that are by no means links or also `javascript:` links which we might want to forbid

mhsdesign committed:
FEATURE: LinkEditor empty links with `anchor` to target current page like `#section-a`

This has been asked for in the Archeopteryx package which actually resulted in the introduction of the CustomLink: https://github.com/sitegeist/Sitegeist.Archaeopteryx/commit/0ba0940ad69113b187374c85c684d491eba814f9

As link options like anchor are separated from the link type model we have to allow empty links like `#`

mhsdesign committed:
WIP: LinkEditor simplify Web dialog by allowing all kinds of links and protocols and also include the `anchor` option as plain text

this allows insert urls like

```
seb://bist-du-glücklich?ja=true#test
```

Instead of hard validation errors we now emit warnings as in that the apply button is not blocked.

To allow determining the anchor we disable the anchor text field option as there is no two-way-binding in the LinkType framework yet and its questionable if its sensible.

mhsdesign committed:
WIP: LinkEditor emit warnings as tooltips and reintroduce old format as http link option for web link type

mhsdesign committed:
PATCH: useLatestValueFrom.ts only call `setValue` when value changed

refs 695bc926df7d4649f7be736a8fc2cc52d6843a61

mhsdesign committed:
TASK: Introduce utility to pick state$ `pick()`

this allows us to store all current link models in a single observable and pick from it to access to pass it to the desired component

mhsdesign committed:
PATCH: Improve state pick() utility

mhsdesign committed:
TASK: Rename `createActionObservable` to `createChannel`

The primitive is a channel not an event or an action as this is dispatched through the thing now named "channel"

mhsdesign committed:
!!! FEATURE: Simplify link editor by not separating anchor from href and displaying it always separately

With the configuration `anchor: true` an editor option for the anchor was previously enabled. This was necessary in the old link editor to enable anchors on node urls. The old link editor though did allow inserting an anchor into the manual link field either way - enabling the anchor field would allow to edit this segment separately with a two-way binding. In the archaeopteryx the pattern of having to enable anchors was continued and the web tab forced to specify a base url and did not allow relative anchors until recently.

A recent hack now allowed to edit the full url with anchor in the web tab by returning in `convertModelToLink` also the `options.anchor` just to have that in the inspector glued back together again. This clearly breaks the abstraction, and it became imminent that the anchor is part of the href and should not be treated as separate option.

The anchor option is now displayed separately for the document and asset tab. The web tab will allow the anchor to be set in the main field together with the rest of the link.

Breaking changes:
----------------

The global editor option for the inspector and for the inline link editor `anchor: <boolean>` is no longer respected.

Instead, anchors are enabled for the asset and node tab at all times.

mhsdesign committed:
WIP: Feature hide advanced link options next to link preview and show all enabled options

During editing, it might be confusing to have switching options per tab at the same locations leading to either missing options per tab, and it makes the layout less familiar and easy to use.

Before selecting a link there is also no need to show these options and even after selecting a link we hide the settings behind a button. If there are changes and the button i collapsed again it will be marked as orange indicating dirty state.

Opening a link with options set will also immediately show all the options to avoid overseeing them when editing. This is especially important when a person edits the link which did not create it to find and identify set options.

mhsdesign committed:
WIP: Introduce utility `mapState` to only select or compute a portion of a state to attempt to reduce re-renders

it might not yield the wanted effect as objects returned will still differ in equality from computation to next computation even though they have the same values to account for that `useLatestValueFrom` now does a simple comparison to assert that first level object keys are equal by reference

mhsdesign committed:
WIP: Merge linkModels into form state and compute formStatus via mapping function

also fix PreviewForLinkType to show the link options as soon as the currently edited link turns valid

e2e pass

mhsdesign committed:
TASK: Extract `shallowEqual` into own module and add tests

react-redux' `useSelector` does also have a comparison

Uses the `exports` pattern to prevent leaking internal things outside

mhsdesign committed:
TASK: Remove now obsolete `isUri` implementation

mhsdesign committed:
TASK: LinkInspector editor respect `disabled` option

mhsdesign committed:
TASK: Remove `anchor: true` from tests

mhsdesign committed:
TASK: Replace `styled-components` with css modules in link editor packages

mhsdesign committed:
TASK: Replace `styled-components` with css modules in link editor core presentation

mhsdesign committed:
PATCH: Solve little typescript riddles

`IconButton` allows `isActive` because of ...rest

and the dialog does not require the `type` to be set.

mhsdesign committed:
WIP: Phone number combine country calling code field with text field and introduce intelligent country formatting for `+`

mhsdesign committed:
TASK: Revert introducing big library `libphonenumber-js` for phone number validation and country calling code suggestions

The library is quite big due to the dataset. Now while it gives some validation when parsing a phone number and also can help by finding the correct country calling code - these features do not justify the complexity of the new library.

We expect that phone numbers are usually inserted as is with or without country code. Its seldom of need for a neos editor to have an overview within neos of all 300+ country calling codes and its not trivial to build good ux for that.

For further simplification phone numbers are now a simple text field with basic warnings if empty or non digits are entered.

mhsdesign committed:
TASK: New linke editor prefer soft validation in mailto tab as well, removing hard constraints

we emit a warning but do not prevent apply because validating a full email is almost impossible :D

the editor should always be in charge of the content!

mhsdesign committed:
PATCH: Ensure that after deleting initial link re-deleting a new link does not reappear initial link

mhsdesign committed:
PATCH: Fix remaining typescript errors in LinkType models

mhsdesign committed:
PATCH: Adjust unit tests to anchor refactoring (anchor is part of href now)

mhsdesign committed:
PATCH: Avoid introducing dependency updates via this change set

`@types/react-redux` leads to a minor update in `redux` to `4.2.1` which we will test separately

`tsx` requires esbuild 25 as dependency while we depend still on 17 this should be updated in sync

mhsdesign committed:
PATCH: Remove little used typescript dev utility `ts-toolbelt`

mhsdesign committed:
PATCH: revert using `main` field instead of exports for now as typescript 4.6 seems too old for this

mhsdesign committed:
TASK: Commit remainding new @types yarn packages

mhsdesign committed:
PATCH: import `PanelProps` from source as `types` is not compiled always

mhsdesign committed:
TASK: Adjust to fixed and new dialog positioning #4007

With "FEATURE: Position Dialog Modals centered and adjust width and height" (see https://github.com/neos/neos-ui/pull/4007)

We no longer require the Modal css hack.

Also, the component is now inlined to simplify the code ;)

mhsdesign committed:
PATCH: Apply eslint's idea how the code should look like

mhsdesign committed:
TASK: Fix remaining eslint errors manually

mhsdesign committed:
PATCH: Adjust e2e to removed Modal component

pKallert commented:
Tested it locally and have some suggestions: 

1. Validate input for anchor (spaces allowed right now) ✅ 
2. The settings button on the upper right when editing links -> Not sure what this is supposed to do, perhaps add a title to indicate what it should do, or why it is disabled? ✅ 

Very Nitpicky suggestions for styling& usability:
4. Add some spacing above anchor  ✅ 
5. When I create & copy to another language and the document does not exist, it says [deleted document], but should correctly say something like "document does not exist" ✅ 
6. When applying filter it says "loading", perhaps some spacing there? ✅ 
7. Assets tab: Can we place the "anchor" outside the scrollable space?  ✅ 

Had a look at the code and it looks good, but also not did not check that deeply since it is a lot of changes 
Very nice new feature 👍

mhsdesign committed:
PATCH: Compatbility to behat/gherkin v 4.14.0 where examples must come at the end

[Behat\Gherkin\Exception\UnexpectedParserNodeException]

Expected Examples table or end of Scenario, but got Step

Commit:
PATCH: adjust text fixtures to expectations

lorenzulrich commented:
For your consideration: I raised some minor issues with Archaeopterix that could maybe be checked/implemented while this is not merged yet:

* Naming of the "Document" tab: https://github.com/sitegeist/Sitegeist.Archaeopteryx/issues/53
* Showing the current link only in the relevant tab (or better MHS's suggestion - maybe this is already done): https://github.com/sitegeist/Sitegeist.Archaeopteryx/issues/49

mhsdesign committed:
WIP vertical tabs linkeditor

mhsdesign committed:
TASK: Remove now unused `@neos-project/utils-helpers` `isEmail` helper in favour of `isemail`

the library was introduced with https://github.com/neos/neos-ui/pull/2988 and handles more cases. A regex would probably be faster and less bloat but as we have the lib already.

For the MailTo field a simple stupid validator like `a-z@a-z.a-z` would probably do already as the worst thing is that a warning is emitted - we dont block the change. And if some redactor enters an email and forgets the tld its probably a mistake and one is not running on a local server with `mail@server1` - but as said it removed things for now so :)

mhsdesign committed:
FEATURE: Group href and link options in "advanced options" details below

This is the third approach to where the options could reside. Initially they were below with anchor and title and rel mixed. Then we - correctly - separated anchors from the options as in `Web` we want to allow to insert a simple anchor directly. And we moved the other link options to the top next to the preview as we wanted to symbolize that they are really available everywhere for every link - and that they affect the set link and are not dependent on the tab. But in the dresden sprint it was found that the anchor as well as the others are really "Advanced" in some common way so we want to group and hide them together below in every tab.

mhsdesign committed:
PATCH: Fix menu label breaking because of space as the width is not claimed

mhsdesign committed:
TASK: Always validate phone model - also when reopening

mhsdesign committed:
TASK: Prevent content shift when selecting new asset by providing loading preview which is equally height as the image

mhsdesign committed:
TASK: Prevent content shift in `SelectNodeTypeFilter` when options are not loaded yet

We add an empty entry to prevent content shift that way the dropdown icon is added as at least one option exists.

The selectbox is disabled during this phase either way so this entry is never visible.

mhsdesign committed:
TASK: Prevent content shift in MailTo because of body text area

The auto size text are doesn't define a good min height itself

But we probably want this default behaviour in other places like the inspector?

mhsdesign committed:
TASK: Simplify "Phone Number" label to "Phone" so entries are similar wide

mhsdesign committed:
TASK: Ensure (for vertical tabs) that full label of tab is shown and label is not placed on second line

mhsdesign committed:
TASK: Show advanced options when opening link editor if they some are set

mhsdesign committed:
TASK: Style advanced button and opened area

mhsdesign committed:
PATCH: Prevent layout shift due to `disabled` `SelectNodeTypeFilter`

The `TextArea` gets a `disabled` attribute which is the problem - safari could not tell me why

mhsdesign committed:
PATCH: Show chevron for advanced toggle state

mhsdesign committed:
TASK: Unify styling of labels

No following colon (this is not UI design)
Checkbox with label on same line

mhsdesign committed:
TASK: Fix label if document node does not exist

previously it showed "Deleted Node" but that is not necessarily true. If a node was copied to another dimension where the link target does not exist the label is confusing. This one is more generic.

mhsdesign committed:
PATCH: Align `Deletable` to use brand hover style and X like other inspector editors to clear input

mhsdesign committed:
PATCH: Add proper loading spinner for NodeTree

mhsdesign committed:
PATCH: Vertical tabs dont show left border twice and prevent `margin-bottom: -1px;` style

mhsdesign committed:
TASK: Dont use button but custom component so text hover only highlights the text to be less bloated

THis effect is similar to the edit preview mode selection.

mhsdesign committed:
TASK: Simplify Link Editor framework by removing first class support for asynchronous models

There is a lot of complexity to async  `convertLinkToModel` because we cannot initially merge this value into the linkTypes observeable state -> this was not a problem as all `Editor` work with a nullable `State` either way as its would also be null if not loaded.

But we also want to open the additional link options if the link has `anchor` set for example. For that we need the resolved link model see linktype.isValid and we were only be able to compute this one tick later. That creates glitches and content shifts for the opening animation of the dialog.

Technically we could solve that by strict separating `DialogWithValue` into `DialogWithValue` and `DialogWithAsyncValue` where the latter would receive a promise and use `usePromise` to resolve that. Splitting it this way would allow non async values to appear immediately which is the default case and async values to be displayed after a loading animation.

Now as the core does not need any async link editors we dont need to add this complexity right now just to carry it around. In case a plugin wants to register a dialog with async state it would need to simplify manage that itself.

mhsdesign commented:
Current state of things, thanks for all the feedback from various occasions:

<img width="911" height="598" alt="image" src="https://github.com/user-attachments/assets/77e0d4a5-5233-455f-9f5b-0a032b60cc2e" />

Alle the things

<img width="1040" height="953" alt="image" src="https://github.com/user-attachments/assets/0cc0a44a-20ae-4604-a1d2-4182bf8c4e3d" />

We now have vertical tabs which fits overall better as it was odd that the preview of the links was above the tabs which made a weird separation.
Further the advanced options are now collapsed below

mhsdesign committed:
PATCH: Remove `Layout.Columns` which set hard coded `min-width: 600px;`

THe MailTo should also not use columns but one entry after the other -> columns didnt work here via the `Layout.Columns` so no difference.

mhsdesign committed:
PATCH: Remove obsolete `LinkToArrayForNeosUiConverter` hack

the fix https://github.com/neos/neos-ui/pull/4022 is merged in the core;)

mhsdesign committed:
PATCH: Remove obsolete `{gridColumn: '1 / -1'}` styling for now flat MailTo fields

mhsdesign committed:
PATCH: Extract inline style to css module if sensible

mhsdesign committed:
TASK: Use `#[Flow\Route]` instead of unused `'@action': 'ignored'` and deprecated `@subpackage`

This hopefully makes christian happy.

We place the controllers to handle the endpoints into the `LinkEditor` namespace. To archive that on normally has to use `@subpackage` to descend into the namespace. We dont want to promote that behaviour and instead want to make controllers discoverable in all places and identify them simply by FQN.
Also for these query endpoints we currently didnt use the ActionController but simply the `ControllerInterface` and handled the `ActionRequest` to a single `processQuery`. This means we ignored the `@action` route value. To make that transparent we wrote `@ignored` into that - by default it would just be `index`. This also seems we work more around the framework and we dont want to look like we do :) -> in the future flow should provide a way to legally omit the `@action` and have it default to `processRequest` or something.

As these things are not implemented in Flow we make use of the `#[Flow\Route]` attribute which already as part of forwards compatibility allows to place controllers quite flexible (it will still use the @subpackage bts) but its part of the core and allowed to be used this way.
Also we now use the @action routing again but we just simply by convention only use one per controller.

The `AbstractQueryController` behaves now like and `ActionController` but doesnt invoke all the crazy property mapping and things we dont want to use :)

mhsdesign committed:
TASK: Add translation keys for web link validation

mhsdesign committed:
TASK: Add basic validation to node and asset anchor that its not only spaces

mhsdesign committed:
PATCH: Make eslint happy and remove todo that we want url decoding

Url decoding in javascript will always end up double encoding the urls incrementally

mhsdesign committed:
TASK: Add typescript typing for registries containers, inspector, richtextToolbar, ckEditor5

mhsdesign committed:
PATCH: Add `tabs--vertical` ts type and update snapshot

Sebobo committed:
BUGFIX: Use Neos label components for linkeditor

Sebobo committed:
BUGFIX: Make tab buttons full width in vertical layout

Sebobo committed:
BUGFIX: Correct linkbutton icon

Sebobo committed:
TASK: Remove unused container component

Sebobo committed:
TASK: Use flex for stack layout

Sebobo committed:
TASK: Cleanup advanced button classes

Sebobo committed:
TASK: Add label for advanced button

Sebobo committed:
TASK: Cleanup indentation

mhsdesign committed:
TASK: Inline `makeLinkType()` to simplify LinkType instantiation and registration

`id` is now available in the module instead of repassed as argument

Sebobo committed:
BUGFIX: Advanced button width

mhsdesign committed:
TASK: Prefer conventional jest tests over not run native node tests

The `Sitegeist.Archaeopteryx` went with https://github.com/sitegeist/Sitegeist.Archaeopteryx/pull/83 in the direction to abandon jest. We also might plan that for Neos as jest was seemingly dead and brings a lot of ugly dev dependencies like babel.
This link editor change also introduced this new node test running but it was commented out as the new dependencies need to be organised first. That leaves us with not run `test.js` files. Now we reevaluated that we probably have to keep jest alive for some time still and thus all the new tests are now adjusted to jest style.

Sebobo committed:
TASK: Use spacing variables for linkeditor

Sebobo committed:
FEATURE: Vertical tabnav can be sticky

Sebobo committed:
BUGFIX: Prevent scrollbar issues with vertical tabs in dialog

mhsdesign committed:
TASK: Adjust link editor e2e to current state - use fixed ids in the pattern or `#neos-LinkEditor-...`

mhsdesign committed:
TASK: Adjust to Neos core Link value object (https://github.com/neos/neos-development-collection/pull/5682)

mhsdesign committed:
PATCH: Adjust eslint, jest snapshots and fix rare FF 143 condition in `formattingOptions`

The web tab renders "formattingOptions", __get$(...) is undefined" in the e2e tests in CI

Sebobo committed:
BUGFIX: Prevent scrollbar when advanced section is collapsed

Sebobo committed:
BUGFIX: Remove unused ref that triggers React warning

Sebobo committed:
BUGFIX: Make dnd type for nodetree optional to prevent warning

Sebobo committed:
FEATURE: Move advanced link editor to bottom left of dialog

Sebobo committed:
TASK: Improve layout in asset link selection

Sebobo committed:
TASK: Improve linkeditor button styling

Sebobo committed:
TASK: Add internal note to new tabs props

Sebobo reviewed (approved):
Best PR ⚡️

mhsdesign committed:
TASK: Adjust naming of inspector and inline link editor adapter packages

mhsdesign committed:
TASK: Remove semi namespace `LinkEditor:` from `LinkEditor:Node` instead allow to configure as `linkTypes.Node` instead

mhsdesign committed:
TASK: Rename `PhoneNumber` to `Phone` in code and also adjust id to `Phone`

mhsdesign committed:
PATCH: Use `position` api for `linkTypes` to make `Node` (document) tab the first for e2e testing

mhsdesign committed:
PATCH: Move react components to separate files to separate style css module usage

mhsdesign committed:
PATCH: Dont wrap editor in `Layout.Stack` from outside but only where applicable from inside

mhsdesign committed:
PATCH: Remove fixed height to increase size of media iframe browser

mhsdesign committed:
PATCH: Remove obsolete `sticky` state as we scroll the tab contents now

Sebobo committed:
BUGFIX: Prevent second scrollbar with asset selection in linkeditor

mhsdesign committed:
BREAKING NEWS: Eslint config turns developer crazy - and he jumps out of a window

... but he cant because the window contains a linting issue and thus the window did not compile and crashed during opening

end of story

mhsdesign committed:
TASK: Overhaul dialog title and apply button labels

Instead of a varying label for Create {linkType} Link - which is hard to translate for example in german and still looks odd if quoted "Dokument"-Link erstellen - we have now a static title based on if there was an initial link or not.

We also change the label of the button if a link is updated or newly created - this should allow a more clear workflow as even using a new tab we still update a link and not as before create a new one. (Same with delete)
